### PR TITLE
Add new p3p solver

### PR DESCRIPTION
**Pull request by yaqding**
Slightly more efficient than existing methods.

### CODE REVIEW

**Comment by pmoulon** [comment]
Thank you for the contribution. Let me test it on various dataset and come back to you.

File: src/openMVG/multiview/solver_resection_p3p_ding.cpp [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+      return;
+    double x11 = lambda1 - lambda2 * b12;
+    double x12 = lambda2 - lambda1 * b12;
+    double x21 = lambda1 - lambda3 * b13;
+    double x23 = lambda3 - lambda1 * b13;
+    double x32 = lambda2 - lambda3 * b23;
+    double x33 = lambda3 - lambda2 * b23;
+    double detJ = 0.5 / (x11 * x23 * x32 +
+                         x12 * x21 * x33); // half minus inverse determinant
+    // This uses the closed form of the inverse for the jacobean.
```
Comment (pmoulon) [inline src/openMVG/multiview/solver_resection_p3p_ding.cpp]: jacobean -> jacobian

File: src/openMVG/multiview/solver_resection_p3p_ding.cpp [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+    // Due to the zero elements this actually becomes quite nice.
+    lambda1 += (-x23 * x32 * r1 - x12 * x33 * r2 + x12 * x23 * r3) * detJ;
+    lambda2 += (-x21 * x33 * r1 + x11 * x33 * r2 - x11 * x23 * r3) * detJ;
+    lambda3 += (x21 * x32 * r1 - x11 * x32 * r2 - x12 * x21 * r3) * detJ;
+  }
+};
+
+static inline bool root2real(const double b, const double c, double &r1,
```
Comment (pmoulon) [inline src/openMVG/multiview/solver_resection_p3p_ding.cpp]: Some code is most likely identical to https://github.com/openMVG/openMVG/blob/develop/src/openMVG/multiview/solver_resection_p3p_nordberg.cpp#L96
We could maybe factorize in a `p3p_nordberg_utils.hpp` file

Comment (yaqding) [reply src/openMVG/multiview/solver_resection_p3p_ding.cpp]: Moved the repeated scripts into one file.

**Comment by pmoulon** [comment]
Starting to test this now, and results seems good.
Need to compare some reconstruction statistics side by side and I will merge the PR ;-)

**Comment by pmoulon** [comment]
Thank you for your contribution @yaqding 
I tested on multiple scene and it provide very close performance, or identical to the previous p3p solver, and seems to be slightly faster.

**Pull request merged**